ShibbolethWAYF support

xrosecky opened the pull request:
ShibbolethWAYF implementation (where are you from) - support of multiple IdPs for authentication for use in combination with MultiBackend driver. 

- Common functionality moved to AbstractShibboleth
- Discovery service with WAYF is required - see [IdPDiscovery](https://wiki.shibboleth.net/confluence/display/CONCEPT/IdPDiscovery).
- Mapping of attributes can be overriden for each IdP

xrosecky committed:
ShibbolethWAYF support

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -69,230 +66,16 @@ public function __construct(\Laminas\Session\ManagerInterface $sessionManager)
+     * @return array Shibboleth configuration to use
      */
-    protected function getRequiredAttributes()
+    public function getShibbolethConfiguration($request)
```
Should this be protected?

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,345 @@
+        if (null === $shibSessionId) {
+            return;
+        }
+        $localSessionId = $this->sessionManager->getId();
```
The sessionManager property is used here in AbstractShibboleth, but it is only defined and populated in the regular Shibboleth class; I assume this is going to cause problems in ShibbolethWithWAYF. You might want to review this part of the refactoring.

demiankatz reviewed (changes requested):
I'm not currently in a good position to test this, but I did notice a couple of possible issues in a quick review. Once that's cleaned up, it might be worth asking for a review from @EreMaijala when he returns to the office (I believe he is on leave at the moment).

xrosecky commented:
OK, I have fixed the mentioned issues.

I am also working on shibboleth support for library cards.

demiankatz commented on config/vufind/shibboleth.ini:
```diff
@@ -0,0 +1,11 @@
+; Each section of the file contains IdP configuration.
```
Do you think it would make sense to rename this ShibbolethWithWAYF.ini so it matches up with the authentication plugin? That way it is less likely to confuse users of the regular Shibboleth plugin.

xrosecky replied:
Yes, it makes more sense, renamed.

demiankatz commented:
Thanks, @xrosecky, this is looking good. I had one more small thought which I have commented on above. After that, please let me know if you'd like me to do a closer review of the code with an eye to merging soon, or if you'd prefer to wait and make more progress on the barcode work (and/or wait for input from @EreMaijala when he is back).

xrosecky committed:
configuration file for ShibbolethWithWAYF renamed

xrosecky commented:
I will wait for input from @EreMaijala. In the meanwhile I will start working on the library card support with Shibboleth.

demiankatz commented:
Thanks, @xrosecky! I've requested a review from @EreMaijala, so he'll see this in his queue when he returns.

EreMaijala commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,362 @@
+                } elseif ($attribute == 'cat_username' && isset($shib['prefix'])) {
+                    $user->cat_username = $shib['prefix'] . '.'
+                        . (($value === null) ? '' : $value);
+                } elseif ($attribute != 'cat_password') {
```
I'd switch this around and check for `} elseif ($attribute === 'cat_password') {` here and handle other values in the final else block. I know it's just a copy from old code, but it just bugs me. Feel free to ignore if you'd rather keep the moved code intact. :)

EreMaijala commented on config/vufind/ShibbolethWithWAYF.ini:
```diff
@@ -0,0 +1,11 @@
+; Each section of the file contains IdP configuration.
+; The section name is important, because defines prefix used for creating user cat_username attribute, that is used
```
Should this be "because *it* defines"?

EreMaijala commented on module/VuFind/src/VuFind/Auth/ShibbolethWithWAYF.php:
```diff
@@ -0,0 +1,106 @@
+ */
+class ShibbolethWithWAYF extends AbstractShibboleth
+{
+    protected $shibbolethConfig;
```
Needs documentation.

EreMaijala reviewed (changes requested):
All in all this looks good to me. 

We actually have a local customization for mapping a Shibboleth attribute to cat_username for per entity ID, but this is a much cleaner solution with the ability to customize other attribute mappings too.

xrosecky committed:
if condition simplified

xrosecky committed:
return null for missing attribute

xrosecky committed:
typo in documentation fixed

xrosecky committed:
documentation added

demiankatz commented on module/VuFind/src/VuFind/Auth/ShibbolethWithWAYF.php:
```diff
@@ -41,6 +42,12 @@
+    /**
+     * Configured IdPs with entityId and overridden attribute mapping
+     *
+     * @param \Laminas\Config\Config
```
I think the Travis build is complaining because this should be `@var` rather than `@param`.

xrosecky committed:
code style issues fixed

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,362 @@
+<?php
+/**
+ * Shibboleth with WAYF authentication module.
```
This comment (and the one above the class itself) should probably be updated to say "Abstract Shibboleth base class" or something similar.

demiankatz replied:
This comment still needs to be corrected.

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,362 @@
+        foreach ($shib as $key => $value) {
+            if (preg_match("/userattribute_[0-9]{1,}/", $key)) {
+                $valueKey = 'userattribute_value_' . substr($key, 14);
+                $sortedUserAttributes[$value] = $shib['$valueKey'] ?? null;
```
This doesn't look right -- putting single quotes around `$valueKey` means that the literal string will be used rather than the value of the variable. Is there even a reason that this had to be changed from object notation to array notation in this situation?

demiankatz replied:
...or is the idea that you were supposed to be using the `$config` parameter to this function (which currently appears to be unused) instead of calling `$this->getConfig()`?

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,362 @@
+        if (!isset($shib['session_id'])) {
+            return;
+        }
+        $shibSessionId = getAttribute($request, $shib['session_id']);
```
Is this supposed to be `$this->getAttribute()` rather than just `getAttribute()`?

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,362 @@
+ */
+abstract class AbstractShibboleth extends AbstractBase
+{
+    const DEFAULT_IDPSERVERPARAM = 'Shib-Identity-Provider';
```
It might be a good idea to put some comments on these constants, just to explain their purpose.

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,362 @@
+     */
+    public function isExpired()
+    {
+        $proxy = $this->getConfig()->Shibboleth->proxy ?? false;
```
I see new proxy functionality in a few places, but it does not appear to be documented in config.ini. Does something need to be added or removed?

demiankatz reviewed (changes requested):
I did a more thorough review of this and found a few likely problems. Please see comments below.

The fact that some of these refactoring problems went unnoticed suggests that we need to make our unit tests for this class more robust. If you have time to do that, it would be a great help; if not, let me know and perhaps I can help (though it would be easier if I had write access to this branch).

xrosecky commented:
Thank you for your review, I will fix the issues and add unit / integration tests.

xrosecky committed:
documentation added

xrosecky committed:
configuration for shibboleth proxy + documentation

xrosecky committed:
code fixed

demiankatz commented:
Thanks, @xrosecky, I see you've made some progress. Would you like me to review again, or do you prefer that I wait until after testing is ready?

xrosecky committed:
WIP on integration tests for ShibbolethWithWAYFTest

xrosecky committed:
WIP on integration tests for ShibbolethWithWAYFTest and Shibboleth

xrosecky committed:
WIP on integration tests for ShibbolethWithWAYFTest and Shibboleth

xrosecky committed:
code style fix

xrosecky committed:
more tests for ShibbolethWithWAYFTest

xrosecky committed:
integration tests for ShibbolethWithWAYF finished

xrosecky commented:
OK, I finished tests for ShibbolethWithWAYF, can you review it?

demiankatz commented:
Thanks, @xrosecky! I'm running out of time for reviews this week since the University is closed tomorrow, but I will take a closer look at this early next week. I appreciate all the work on it!

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,372 @@
+            . urlencode($append . 'auth_method=Shibboleth');
+
+        if (isset($config->Shibboleth->provider_id)) {
+            $sessionInitiator = $sessionInitiator . '&entityID=' .
```
Very minor point, but might as well use `.=` here... (I realize this is a pre-existing problem, but might as well fix it while we're refactoring everything).

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,372 @@
+    public function isExpired()
+    {
+        $proxy = $this->getConfig()->Shibboleth->proxy ?? false;
+        return ($proxy) ? isset($_SERVER[$this->normalize(self::SHIB_SESSION_ID)])
```
Should we pass the request object into isExpired() via checkForExpiredCredentials() so we can avoid directly accessing superglobals here? Would that enable us to use the standard getAttribute() support method here instead?

Also, I notice that the current code before refactoring does an `empty` check rather than an `isset` check. Is it better to do that more forgiving check, or is the change done here for a reason?

demiankatz commented on module/VuFind/src/VuFind/Auth/AbstractShibboleth.php:
```diff
@@ -0,0 +1,372 @@
+            $header = $request->getHeader($this->normalize($attribute));
+            return ($header) ? $header->getFieldValue() : null;
+        } else {
+            return $request->getServer()->get($attribute) ?? null;
```
I don't think the `?? null` is needed on this line, since `get` will return null to begin with.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -44,255 +42,23 @@
-     *
-     * @return bool|string
+     * @throws \VuFind\Exception\Auth
+     * @return array Shibboleth configuration to use
```
We should add:

<pre>
     *
     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
</pre>

to the bottom of this comment block to suppress warnings about the intentionally unused formal parameter here.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Auth/ShibbolethWithWAYFTest.php:
```diff
@@ -0,0 +1,237 @@
+        return $request;
+    }
+
+    public function testLogin1()
```
Would be helpful to have doc comments on the tests explaining their purpose, for future reference.

demiankatz reviewed (commented):
Thanks again, @xrosecky; I've given this another fairly close review and left a few more comments. I'll do some hands-on testing once these are resolved. If you do not have time right now, let me know and I can make the edits myself. If you can give me write access to this branch, I can take care of them here as well... but right now I don't have write access, so I can only comment.

EreMaijala commented on module/VuFind/src/VuFind/Auth/ShibbolethWithWAYF.php:
```diff
@@ -0,0 +1,115 @@
+     *
+     * @return string entityId of IdP
+     */
+    protected function fetchCurrentEntityId($request)
```
A minor nit: "fetch" sounds like it will access some external system to fetch the attribute. "get" would be more obvious to me.

demiankatz replied:
Agreed.

EreMaijala commented:
I gave this another thought, and I have one remaining question: Do we actually need ShibbolethWithWAYF as a separate authenticator or could we roll all the functionality into the Shibboleth class? With the separate classes I can see a couple of issues:

1.) ShibbolethWithWAYF will become a distinct authentication method, which means it is separate from Shibboleth even if it's practically the same. This means e.g. reconfiguration of permissions, migration of the user table (auth_method column) would be needed to switch to ShibbolethWithWAYF.

2.) ShibbolethWithWAYF is not really obviously clear as a name. To me it implies that VuFind would do some work of the WAYF or something, which is not really the case.

Both of these can be fixed while keeping the separate class, but I can't really see the need for the split. We could just as well have a Shibboleth.ini that would allow to override IdP specific settings even if only a single IdP is used. I'm thinking of a shared installation (like ours) where we have a central main configuration and instance-specific details. While a single instance typically only uses a Shibboleth IdP directly, we could still benefit from sharing the IdP-specific configuration with all the instances.

demiankatz commented:
Thanks, @EreMaijala, I agree that the name isn't ideal, and it might be preferable to have a single configurable class rather than multiple independent authentication methods. Since all of the customizations here revolves around configuration loading, what if we made a VuFind\Auth\Shibboleth\ConfigurationLoadingInterface, and then injected different implementations of that interface into the Shibboleth handler through its factory? I think that would accomplish the same effect, and leave the door open to other implementations, without creating different named authentication handlers.

xrosecky commented:
> Thanks, @EreMaijala, I agree that the name isn't ideal, and it might be preferable to have a single configurable class rather than multiple independent authentication methods. Since all of the customizations here revolves around configuration loading, what if we made a VuFind\Auth\Shibboleth\ConfigurationLoadingInterface, and then injected different implementations of that interface into the Shibboleth handler through its factory? I think that would accomplish the same effect, and leave the door open to other implementations, without creating different named authentication handlers.

I think that this is a great idea, so I will refactor the code.

xrosecky committed:
WIP on shibboleth refactoring - use ConfigurationLoadingInterface

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth/Shibboleth.php:
```diff
@@ -1,6 +1,6 @@
 <?php
 /**
- * Shibboleth authentication module.
+ * Shibboleth with WAYF authentication module.
```
We should straighten out the comments here.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth/Shibboleth.php:
```diff
@@ -27,15 +27,18 @@
 
 /**
- * Shibboleth authentication module.
+ * Abstract Shibboleth class
```
Another comment that should probably be readjusted.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth/Shibboleth.php:
```diff
@@ -27,15 +27,18 @@
  * @link     https://vufind.org Main Page
  */
-namespace VuFind\Auth;
+namespace VuFind\Auth\Shibboleth;
```
Is it necessary to move everything into the Shibboleth namespace? If we keep the file where it was but add support files within the Shibboleth namespace, I think that would be equally valid, but slightly less disruptive.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth/Shibboleth.php:
```diff
@@ -44,28 +47,125 @@
+     *
+     * @return bool|string
+     */
+    public function getSessionInitiator($target)
```
I see that the order of some methods has been moved around in the file. It might be better to keep the existing method order, because it will make the diffs easier to review in the pull request.

demiankatz commented on config/vufind/config.ini:
```diff
@@ -394,6 +394,7 @@ force_first_scheduled_email = false
 method         = Database
 ;method         = AlmaDatabase
 ;method         = Shibboleth
+;method         = ShibbolethWithWAYF
```
Looks like config.ini needs to be updated to reflect the latest changes -- removing this new method and instead adding the override setting.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth/ShibbolethFactory.php:
```diff
@@ -60,8 +62,17 @@ public function __invoke(ContainerInterface $container, $requestedName,
+        $config = $container->get(\VuFind\Config\PluginManager::class)->get('config');
+        $override = $config->Shibboleth->override;
+        $loader = null;
+        if (!empty($override)) {
```
Might be worth moving this logic to a protected getConfigurationLoader() method of the factory, so it's easier to override with a custom loader by extending the factory.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth/SingleIdPConfigurationLoading.php:
```diff
@@ -0,0 +1,35 @@
+<?php
+namespace VuFind\Auth\Shibboleth;
+
+class SingleIdPConfigurationLoading implements ConfigurationLoadingInterface
```
I would suggest replacing "Loading" with "Loader" in the class and interface names; most of our classes are named for what they do, and I think "Loader" would read more consistently.

demiankatz reviewed (commented):
Thanks for the progress on this; it's looking good! I've left a few comments, some of them rather picky. As always, I can help with some of these changes myself if you're willing to share write access; I'm not trying to make unnecessary work for you, but I'm doing my best to ensure that the PR is as easy to review as possible, so I don't miss anything important!

xrosecky committed:
WIP on shibboleth refactoring - code style issues fixed

xrosecky commented:
@demiankatz thanks for your thorough review, I addressed some of your notes and added write access to our repository for you. I have yet to merge integration tests to one class + cleanup.

demiankatz commented:
Thanks, @xrosecky, I appreciate the write access! In future, if I see very small issues, this will enable me to fix them without having to bother you with an extra cycle of work. :-)

How important is it to you that we get this merged into VuFind 7.0? I think I'm running out of time to finish this up in time for that release, so it's probably safer to target this for 7.1 instead (especially if we want more input from @EreMaijala, who is going to be on vacation for several weeks starting tomorrow). However, if there is an urgent need to finish the work, I can do my best to expedite the remaining review.

xrosecky committed:
shibboleth tests merged

demiankatz commented:
@xrosecky, thanks for the latest progress. Would you like some help cleaning up the style issues Travis is currently complaining about? If so, I can assist when time permits!

Also, do you mind if I rearrange the function order in Shibboleth.php to make the diffs easier to read? If your IDE is rearranging the functions automatically, though, that may cause more problems than it solves. Please let me know your preference!

xrosecky committed:
code style issues fixed

xrosecky committed:
methods reordered to make diff easier to read

xrosecky commented:
I fixed the comments + reordered methods in Shibboleth.php.

demiankatz commented:
This is great, @xrosecky, thank you! I'm nearly finished deploying Shibboleth at my institution, so I'll try to test this in my staging server when it is practical to do so (perhaps in a couple of weeks). I'll also be interested to hear if @EreMaijala has any further thoughts when he returns from his current holiday.

xrosecky committed:
typo fixed - bad variable name, documentation updated

demiankatz committed:
Remove redundant fallback.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -228,18 +250,9 @@ public function getSessionInitiator($target)
-        }
-        return false;
+        $proxy = $this->getConfig()->Shibboleth->proxy ?? false;
+        return ($proxy) ? isset($_SERVER[$this->normalize(self::SHIB_SESSION_ID)])
```
Are you sure these changes are correct/appropriate? I see a few potential issues:

1.) This method is checking to see if the login is expired, but it's checking to see if a session ID is set. Does the logic need to be reversed to check to see if things are NOT set?

2.) It would be better to avoid using PHP superglobals. Can you inject the framework's server object through the constructor so you have access to it here? (You can retrieve it with `$container->get('Request')->getServer()` in the factory). Can you use `getenv()` instead of `$_ENV`?

xrosecky replied:
I fixed the code and added documentation that was removed by mistake.

demiankatz reviewed (changes requested):
This PR hasn't moved forward in a little while, so I gave it another review. It looks like most of the previous comments have been addressed, so I resolved them. One small suggestion to remove a redundant fallback had not been taken, so I went ahead and adjusted it (see commit 9eba596). I think beyond that, there's only one possible outstanding problem.

@xrosecky, could you take a look? I'm also still interested in hearing from @EreMaijala if he has time/interest for one last review.

xrosecky committed:
isExpired in Shibboleth fixed

demiankatz commented:
Thanks for the clarification and fix, @xrosecky! I hope to have time to do some hands-on testing of this code fairly soon; I'll let you know if I encounter any further problems at that time.

demiankatz committed:
Revert unnecessary change.

demiankatz committed:
Merge branch 'dev' into shibbolethWithWAYF

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -228,18 +250,12 @@ public function getSessionInitiator($target)
+        // object... except that the request object doesn't exist yet when
+        // this routine gets called.
+        return ($proxy) ? !isset($_SERVER[$this->normalize(self::SHIB_SESSION_ID)])
+            : !isset($_ENV[self::SHIB_SESSION_ID]);
```
I think `$_SERVER` and `$_ENV` are inverted here -- if proxy is true, we should look at the environment rather than the server, correct? Also, isn't normalization of keys normally associated with the environment, not the server?

xrosecky replied:
It is correct, but I noticed, that the logging of attributes is not working in proxy mode - fixed in the latest commit. See the comments in config.ini for Shibboleth proxy:

Set to true when shibboleth is running on the proxy side and so we have to read shibboleth attributes from headers instead of server environment variables

Are you running Shibboleth in the web server along with the Vufind or do you use reverse proxy in your environment?

demiankatz replied:
I'm running VuFind and Shibboleth on the same server, so I'm using `$_SERVER` rather than headers for attributes. Thus, my proxy variable is set to false. However, as written, this code appears to be saying "if proxy is true, then look at `$_SERVER`; else look at `$_ENV`..." and that seems to be backwards to me).

xrosecky replied:
And do you use Apache web server with Shibboleth module?

demiankatz replied:
Yes, correct. So when I have an active Shibboleth session, `$_SERVER['Shib-Session-ID']` is set. When I do not, the value is unset. However, as written, this code is currently checking for `$_ENV['Shib-Session-ID']` when proxy is set to false, and for `$_SERVER['Shib_Session_ID']` (with - changed to _) when proxy is set to true, so it fails for me in both situations, because neither of those values are ever set. I still think this looks like a bug, and that the code should be 

<pre>
return ($proxy) ? !isset($_ENV[$this->normalize(self::SHIB_SESSION_ID)])
    : !isset($_SERVER[self::SHIB_SESSION_ID]);
</pre>

...but I apologize if I'm persisting in an incorrect assumption. :-)

demiankatz replied:
I also think it may be possible to use the request object instead of the superglobals here, because as far as I can tell, this method only gets through the manager's `checkForExpiredCredentials()` method, which in turn only gets called from this factory: https://github.com/vufind-org/vufind/blob/dc4677e100ecd786e5bb4b5b434a6ea08025aa9c/module/VuFind/src/VuFind/Auth/ManagerFactory.php#L95 -- I don't see why we couldn't add a `$container->get('Request')` parameter to this method call in this context).

xrosecky replied:
That is strange. Maybe you are using headers instead of environment variables for passing shibboleth attributes in your Apache configuration, which is not recommended:

ShibUseHeaders On

See https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPApacheConfig#NativeSPApacheConfig-AuthConfigOptions for details.

demiankatz replied:
Thanks for your latest adjustments to the code; they look likely to fix my problem. I will test when time permits (hopefully next week if not sooner).

demiankatz replied:
@xrosecky, I see that @EreMaijala has raised some new comments, so I'm going to hold off on testing until you have had a chance to discuss those with him. Please let me know when you're ready for me to test again!

demiankatz reviewed (changes requested):
I finally had an opportunity to try this code in my test environment, and it didn't work -- my session always expired immediately after I logged in. I think it is related to the comment below. Can you review/fix? Once you adjust this, I'll try again. Thanks!

xrosecky committed:
Shibboleth: correct logging of attributes for proxy mode

xrosecky committed:
code style issues fixed

xrosecky committed:
Shibboleth: use Http request object in isExpired

xrosecky committed:
code style issues fixed

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -106,22 +155,32 @@ protected function validateConfig()
+        $entityId = $this->getCurrentEntityId($request);
+        $shib = $this->getConfigurationLoader()->getConfiguration($entityId);
+        $username = $this->getAttribute($request, $shib['username']);
+        $proxy = $this->getConfig()->Shibboleth->proxy ?? false;
```
I think it would be better to make the $proxy flag a member attribute and set it once in constructor. Now it's read in several places separately.

EreMaijala commented on config/vufind/config.ini:
```diff
@@ -746,6 +749,9 @@ database          = mysql://root@localhost/vufind
 ;home_library = HTTP_HOME_LIBRARY
+; You can here override mapping or required attributes for specific IdP here - name of configuration
+; file without .ini prefix
+;override = Shibboleth
```
Is there an actual reason the ini file needs to be configurable? Would it suffice to have something like allow_configuration_override=true/false? Or does there need to be a setting at all? Why isn't Shibboleth.ini always read?

demiankatz replied:
I think this is a matter of backward compatibility. If you only have one IdP, you don't need Shibboleth.ini because the [Shibboleth] section of config.ini handles everything. You only need to load the .ini file if you have multiple IdPs to differentiate between. I do agree that at present, there doesn't seem to be a compelling reason to make the actual filename configurable, since it looks like all of the Shib-related settings will always be in a single file anyway. I'd be in favor of simplifying this to a boolean. I also wouldn't object to moving everything into that file in order to reduce the size of config.ini, but if we did that, we'd have to do more extra for backward compatibility/migration, so perhaps that is better addressed in a future PR.

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -106,22 +155,32 @@ protected function validateConfig()
-        foreach ($this->getRequiredAttributes() as $key => $value) {
-            if (!preg_match('/' . $value . '/', $request->getServer()->get($key))) {
+        foreach ($this->getRequiredAttributes($shib) as $key => $value) {
+            if (!preg_match(
```
Write it like this and it'll fit on a single line:

`            if (!preg_match("/$value/", $this->getAttribute($request, $key))) {`

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -295,4 +342,78 @@ protected function getRequiredAttributes()
+            return;
+        }
+        $localSessionId = $this->sessionManager->getId();
+        $externalSession = $this->getDbTableManager()
```
This fits on a single line.

EreMaijala commented on config/vufind/config.ini:
```diff
@@ -725,6 +725,9 @@ database          = mysql://root@localhost/vufind
 ;userattribute_value_2 = member
+; Set to true when shibboleth is running on the proxy side and so we have to
+; read shibboleth attributes from headers instead of server environment variables
+;proxy                  = true
```
Maybe `use_headers = true/false` and add a note that care must be taken when using this option so that the headers cannot be spoofed?

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -135,19 +194,18 @@ public function authenticate($request)
-                    $user->$attribute = ($value === null) ? '' : $value;
-                } else {
+                } elseif ($attribute == 'cat_username' && isset($shib['prefix'])) {
+                    $user->cat_username = $shib['prefix'] . '.'
```
Should cat_username really be set to just the prefix if $value is null?

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth/MultiIdPConfigurationLoader.php:
```diff
@@ -0,0 +1,90 @@
+            );
+            throw new AuthException('Missing configuration for IdP.');
+        }
+        foreach ($idpConfig as $key => $value) {
```
array_merge?

xrosecky committed:
Shibboleth: proxy as member attribute

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -228,18 +274,14 @@ public function getSessionInitiator($target)
-            return empty($username);
-        }
-        return false;
+        // It would be more proper to call getServer on a Laminas request
```
These changes drew my attention on this method and the logout mechanism. I can see it's ancient, but it looks like it might be something that might cause trouble too. E.g. in our case we only enable Shibboleth in Apache when the url contains `auth_method=Shibboleth`, and even then the actual Shibboleth session is short-lived. I suppose we haven't stumbled on this since we use ChoiceAuth and it doesn't implement isExpired(). So, would you be able to explain the use case or whether this is just some ancient code no longer really needed? If this is still necessary for a certain case, can you check the comment and add a configuration setting that explicitly enables the expiration check? It was previously tied to config containing username and logout settings, but e.g. adding the logout setting enabling this would have seem like a suprising side effect.

xrosecky committed:
Shibboleth: code style issues fixed

xrosecky committed:
Shibboleth: misleading comment removed

xrosecky committed:
Shibboleth: proxy renamed to use_headers

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -295,4 +356,73 @@ protected function getRequiredAttributes()
+     *
+     * @return string header name
+     */
+    protected function normalize($attribute)
```
This method might benefit from a more descriptive name, like `normalizeAttrName`, in case in future we need to normalize something else in a different way.

demiankatz commented on module/VuFind/src/VuFind/Auth/ShibbolethFactory.php:
```diff
@@ -60,8 +64,33 @@ public function __invoke(ContainerInterface $container, $requestedName,
+    {
+        $config = $container->get(\VuFind\Config\PluginManager::class)
+            ->get('config');
+        $override = $config->Shibboleth->override;
```
Might be good to add a `?? null` on this so notices aren't generated if the config is missing.

demiankatz replied:
...but of course if we decide to switch the configuration to a boolean, this will become a non-issue. :-)

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -105,23 +174,30 @@ protected function validateConfig()
     public function authenticate($request)
     {
+        // validate config before authentication
+        $this->validateConfig();
```
The `validateConfig` method should be called automatically whenever the built-in `getConfig` method is called (based on logic set up in the authentication base class). Does the introduction of the configuration loader require us to revisit this in any way? I just want to be sure that `validateConfig` is validating the right thing here, and also checking whether there's a more natural place where we can ensure that validation is occurring automatically whenever it is needed. Let me know if you need me to do any deeper analysis on these questions; it's possible that there is no problem here, but this change led me to raise the questions just to be sure we're doing things correctly.

demiankatz commented:
Thanks for the progress, @xrosecky! It looks like you're still working through @EreMaijala's comments, but I left a few more suggestions and questions in the meantime. Apologies that this is taking so long, but it's a complex problem, and I want to be sure it works for everyone. I appreciate your patience!

xrosecky committed:
Shibboleth: configuration override simplified, always read configuration from shibboleth.ini

xrosecky committed:
Shibboleth: normalize renamed to normalizeAttrName

xrosecky committed:
Shibboleth: check if cat_username is not empty

xrosecky commented:
@demiankatz Did you have opportunity to test this PR?

xrosecky commented:
I have also second thought on the attribute normalization when using HTTP headers instead of environment variables. It seems to differ between web servers so maybe better solution is not to normalize them at all and to expect header names directly in configuration.

https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPAttributeAccess#NativeSPAttributeAccess-RequestHeaders

_When using headers, the main difference is that instead of using the names defined via the mapping process, the application must prefix them with "HTTP\_", and in most tools upcase the rest of the name as well. The specifics vary by tool, and in the case of IIS and ASP.NET are even more bizarre because of serious flaws in IIS' CGI implementation._

demiankatz commented:
@xrosecky, I was waiting for your conversation with @EreMaijala to be resolved before testing again. In particular, it looked like the Oct. 5 comment about logout behavior was still unanswered. If that is taken care of now, I'm certainly willing to test again -- but you might want to double check first to be sure.

Regarding the normalization behavior, I don't have a strong preference, as long as the documentation makes it clear what is required to make things work. I think the existing setup is already confusing, because the `username` and `userattribute_x` fields use un-formatted attribute names, but the `cat_username`/`cat_password`/`email`/etc., etc. use the HTTP_ forms of the values. It seems like it would be less confusing if we always referred to attributes in the same format.

What if we set it up so that normalization is applied if the value does not already start with "HTTP_"? This might provide some degree of backward compatibility while also allowing us to change the configuration to be more consistent/understandable.

demiankatz commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -295,4 +358,73 @@ protected function getRequiredAttributes()
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing
+     * account credentials.
+     *
+     * @return \VuFind\Db\Row\User Object representing logged-in user.
```
This should actually `@return void`, shouldn't it?

xrosecky committed:
Shibboleth: correct return type in documentation for storeShibbolethSession

xrosecky commented:
I think safer approach is to remove attribute normalization by removing method normalizeAttrName and to expect header names in configuration when use_headers is on.

demiankatz commented:
@xrosecky, that sounds reasonable to me. That will retain the existing functionality, and will simply require some configuration changes for people who turn on the new option, correct? As long as the comment before the option clearly explains this, I think it should be fine.

xrosecky committed:
Shibboleth: remove normalization of attribute names

xrosecky commented:
I removed the normalization of attribute names. Attributes with shibboleth session and identity provider id can be overridden in configuration for non-standard situations.

demiankatz commented:
Thanks, @xrosecky, this looks good. I think the only remaining open question is whether isExpired() is behaving the way it should. @EreMaijala expressed concern earlier that in some situations, it may not be appropriate to expire the VuFind session just because the Shibboleth session is no longer available. Do you know the history of the development of this code? Should there be some conditions based on configuration which control when/whether the expiration check is performed?

xrosecky commented:
Yes, I will make it configurable.

berndoberknapp commented:
Note that the current code only can return true if a logout URL is configured. If (single) logout is used, the Shibboleth session should exist at least as long as the VuFind session, otherwise the logout won't work. So the code was specifically designed to handle the situation that the user might have logged out. The change therefore in my opinion is wrong.

berndoberknapp commented:
Just for clarification: Unless (single) logout is used, the Shibboleth session only should be used to setup the VuFind session. The assumption is that the logout URL is configured if the IdP supports (single) logout, so a logout might have be triggered by another SP or the IdP and VuFind might not yet be aware of this. So the Shibboleth session needs to be checked if and only if logout is configured.

xrosecky committed:
Shibboleth: check if logout is configured in isExpired method

xrosecky commented:
Done, method isExpired checks if logout is configured.

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -135,19 +229,20 @@ public function authenticate($request)
+                    && !empty($value)
+                ) {
+                    $user->cat_username = $shib['prefix'] . '.'
+                        . (($value === null) ? '' : $value);
```
This check is no longer necessary since there's the `!empty($value)` check above.

EreMaijala commented on module/VuFind/src/VuFind/Auth/Shibboleth.php:
```diff
@@ -229,17 +305,8 @@ public function getSessionInitiator($target)
-            return empty($username);
-        }
-        return false;
+        $sessionId = $this->getAttribute($this->request, $this->shibSessionId);
```
I still don't think this is correct. The fact that we have a Shibboleth logout URL does not mean that the VuFind session should expire when the Shibboleth session is gone. If you want the Shibboleth session expiration to expire the VuFind session too, I think there should be a separate setting (that defaults to false) to indicate that.

This is somewhat related to the fact that verifying the Shibboleth session for each request is expensive. For a busy server the Shibboleth daemon can become a bottleneck, and I wouldn't recommend using `Require shibboleth` globally (see https://vufind.org/wiki/configuration:shibboleth for more information). This means that the Shibboleth session information is not always available anyway.

xrosecky committed:
Shibboleth: configurable session expiration, disabled by default

xrosecky committed:
Shibboleth: unnecessary check removed

demiankatz committed:
Merge branch 'dev' into shibbolethWithWAYF

demiankatz commented:
Thanks for the long, hard work on this, @xrosecky. I believe that all of the outstanding issues have been resolved, and this is now working correctly for me in my test environment. I'll wait another couple of days in case you or @EreMaijala have any last-minute changes/concerns, but I believe that this is ready to merge now.

demiankatz committed:
Minor tweaks to expiration code.

demiankatz commented:
Following some discussion with @EreMaijala, I made a few minor adjustments to the expiration code (see commit 504387b): I moved the configuration check earlier for efficiency, I changed the default expiration check value to true for backward compatibility, and I added some clarifying comments to config.ini. Please let me know if you have any concerns, @xrosecky!

demiankatz committed:
Fix typo in comment.

demiankatz committed:
Merge branch 'dev' into shibbolethWithWAYF

demiankatz committed:
Merge branch 'dev' into shibbolethWithWAYF

xrosecky commented:
@demiankatz Can you merge this pull request? I have no objections.

demiankatz commented:
Thanks, @xrosecky, I'll merge this sometime this week! (Still catching up on post-holiday tasks, but I should get to this today or tomorrow).
